File: src/main/java/no/nav/syfo/consumer/ws/JournalConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.domain.InngaaendeJournal;
import no.nav.syfo.domain.Inntektsmelding;
import no.nav.syfo.util.JAXB;
import no.nav.tjeneste.virksomhet.journal.v2.*;
import org.springframework.stereotype.Component;

import javax.xml.bind.JAXBElement;

@Component
@Slf4j
public class JournalConsumer {

    private JournalV2 journalV2;

    public JournalConsumer(JournalV2 journalV2) {
        this.journalV2 = journalV2;
    }

    public Inntektsmelding hentInntektsmelding(String journalpostId, InngaaendeJournal inngaaendeJournal) {
        WSHentDokumentRequest request = new WSHentDokumentRequest()
                .withJournalpostId(journalpostId)
                .withDokumentId(inngaaendeJournal.getDokumentId())
                .withVariantformat(new WSVariantformater().withValue("ORIGINAL"));

        try {
            final byte[] inntektsmeldingRAW = journalV2.hentDokument(request).getDokument();
            String inntektsmelding = new String(inntektsmeldingRAW);

            JAXBElement<Object> jaxbInntektsmelding = JAXB.unmarshalInntektsmelding(inntektsmelding);

            XMLInntektsmelding xmlInntektsmelding = jaxbInntektsmelding.getValue() instanceof no.seres.xsd.nav.inntektsmelding_m._20180924.XMLInntektsmeldingM
                    ? InntektsmeldingArbeidsgiver20180924Mapper.INSTANCE.tilXMLInntektsmelding(jaxbInntektsmelding)
                    : InntektsmeldingArbeidsgiverPrivat20181211Mapper.INSTANCE.tilXMLInntektsmelding(jaxbInntektsmelding);

            return new Inntektsmelding(
                    xmlInntektsmelding.getArbeidstakerFnr(),
                    xmlInntektsmelding.getVirksomhetsnummer(),
                    xmlInntektsmelding.getArbeidsgiverPrivat(),
                    journalpostId,
                    xmlInntektsmelding.getArbeidsforholdId(),
                    xmlInntektsmelding.getAarsakTilInnsending(),
                    inngaaendeJournal.getStatus(),
                    xmlInntektsmelding.getPerioder());
        } catch (HentDokumentSikkerhetsbegrensning e) {
            log.error("Feil ved henting av dokument: Sikkerhetsbegrensning!");
            throw new RuntimeException("Feil ved henting av dokument: Sikkerhetsbegrensning!", e);
        } catch (HentDokumentDokumentIkkeFunnet e) {
            log.error("Feil ved henting av dokument: Dokument ikke funnet!");
            throw new RuntimeException("Feil ved henting av journalpost: Dokument ikke funnet!", e);
        } catch (RuntimeException e) {
            log.error("Klarte ikke å hente inntektsmelding med journalpostId: {} og dokumentId: {}", journalpostId, inngaaendeJournal.getDokumentId(), e);
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/ws/JournalConsumer.kt
package no.nav.syfo.consumer.ws

import kotlinx.coroutines.runBlocking
import log
import no.nav.syfo.behandling.HentDokumentFeiletException
import no.nav.syfo.consumer.rest.aktor.AktorConsumer
import no.nav.syfo.consumer.ws.mapping.InntektsmeldingArbeidsgiver20180924Mapper
import no.nav.syfo.consumer.ws.mapping.InntektsmeldingArbeidsgiverPrivat20181211Mapper
import no.nav.syfo.domain.InngaaendeJournal
import no.nav.syfo.domain.JournalStatus
import no.nav.syfo.domain.inntektsmelding.Inntektsmelding
import no.nav.syfo.saf.SafDokumentClient
import no.nav.syfo.saf.SafJournalpostClient
import no.nav.syfo.saf.model.JournalpostResponse
import no.nav.syfo.util.JAXB
import no.seres.xsd.nav.inntektsmelding_m._20180924.XMLInntektsmeldingM
import java.time.LocalDateTime
import javax.xml.bind.JAXBElement

class JournalConsumer(
    private val safDokumentClient: SafDokumentClient,
    private val safJournalpostClient: SafJournalpostClient,
    private val aktorConsumer: AktorConsumer)
{
    var log = log()

    /**
     * 1 - Henter inntektsmelding fra journalpost i byteArray
     * 2 - Gjør om bytearray til XML
     * 2 - mapper om i to separate format
     * 3 - privat mapper henter ut aktørID
     */
    fun hentInntektsmelding(journalpostId: String, arkivReferanse: String): Inntektsmelding {
        try {
            return runBlocking {
                val response: JournalpostResponse = safJournalpostClient.getJournalpostMetadata(journalpostId)!!.data
                val journalpost = response.journalpost
                val inntektsmeldingRAW = safDokumentClient.hentDokument(journalpostId, journalpost.dokumentId!!)
                val jaxbInntektsmelding = JAXB.unmarshalInntektsmelding<JAXBElement<Any>>(inntektsmeldingRAW?.decodeToString())
                val mottattDato: LocalDateTime = journalpost.mottattDato!!
                val journalStatus: JournalStatus = journalpost.journalstatus!!
                if (jaxbInntektsmelding.value is XMLInntektsmeldingM)
                    InntektsmeldingArbeidsgiver20180924Mapper.tilXMLInntektsmelding(jaxbInntektsmelding, journalpostId, mottattDato, journalStatus, arkivReferanse)
                else
                    InntektsmeldingArbeidsgiverPrivat20181211Mapper.tilXMLInntektsmelding(jaxbInntektsmelding, journalpostId, mottattDato, journalStatus, arkivReferanse, aktorConsumer)
            }
        } catch (e: RuntimeException) {
            log.error( "Klarte ikke å hente inntektsmelding med journalpostId: $journalpostId", e )
            throw HentDokumentFeiletException(journalpostId, e)
        }
    }
}
